### PR TITLE
Robust ROCAUC for binary classification. fix #1040

### LINKED ISSUE #1040: ROCAUC treats binary classification as multiclass for estimators with predict_proba available
**Describe the bug**
``ROCAUC`` plots 4 different curves for a binary classification problem when using an estimator with ``predict_proba``, which returns predictions of shape ``nx2``. Only one curve is plotted when estimator only has ``decision_function``, which returns predictions of shape ``nx1``. I think only one ROC curve should be drawn for a binary classification problem, as in the binary case the two ROC curves before averaging are just a reflection of each other.

For comparison, ``PrecisionRecallCurve`` plots 1 curve for binary classification irrespective, whether provided estimator uses ``predict_proba`` or ``decision_function``.

**To Reproduce**

To see this behavior it is enough to look at 
``yellowbrick/tests/baseline_images/test_classifier/test_rocauc/test_binary_decision.png`` and ``yellowbrick/tests/baseline_images/test_classifier/test_rocauc/test_binary_probability.png``

**Expected behavior**

I expected to see just one ROC for binary classification instead of separate curves for class 0 and 1 and their micro and macro averaging.

### PR DESCRIPTION
**Pull request by VladSkripniuk**
Unify the way ROCAUC treats binary classifiers with predict_proba and
decision_function methods.

In this PR I suggest a different approach to fix #1040. The problem which this PR is solving is best explained in this [comment](https://github.com/DistrictDataLabs/yellowbrick/pull/1041#issuecomment-610248840). This PR makes it possible to draw single ROC curve for binary classifiers with `predict_proba` method (which was previously raising an error):
```
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC

from yellowbrick.classifier.prcurve import PrecisionRecallCurve
from yellowbrick.classifier.rocauc import ROCAUC

X, y = load_breast_cancer(return_X_y=True)

X_train, X_test, y_train, y_test = train_test_split(X, y)
viz = ROCAUC(LogisticRegression(), micro=False, macro=False, per_class=False)
viz.fit(X_train, y_train)
viz.score(X_test, y_test)
viz.show()
```
image

I have made the following changes:

1. Changed ROCAUC to treat binary classifiers with predict_proba and decision_function in a unified way.
2. I removed `test_binary_per_class_error` because it's possible to draw 2 ROC curves for `LinearSVC` without raising an error.
3. I removed `test_rocauc_no_classes` because drawing `micro` and `macro` curves when `per_class=False` would make the logic more complicated.
4. I removed `test_rocauc_no_curves` because it's possible to draw a single ROC curve for `LogisticRegression` without raising an error.
### CHECKLIST

- [x] _Is the commit message formatted correctly?_
- [ ] _Have you noted the new functionality/bugfix in the release notes of the next release?_

- [x] _Included a sample plot to visually illustrate your changes?_
- [x] _Do all of your functions and methods have docstrings?_
- [x] _Have you added/updated unit tests where appropriate?_
- [x] _Have you updated the baseline images if necessary?_
- [x] _Have you run the unit tests using `pytest`?_
- [ ] _Is your code style correct (are you using PEP8, pyflakes)?_
- [ ] _Have you documented your new feature/functionality in the docs?_

- [ ] _Have you built the docs using `make html`?_

### CODE REVIEW

**Comment by lwgray** [comment]
@VladSkripniuk We are so appreciative of all your issues and PRs.   I've enjoyed working with you.  However, like most open-source projects we are a very small team that does this in our spare time.  We are all quite busy but we should be able to get your PRs reviewed/merged into our next release (Which is not too far off).

File: yellowbrick/classifier/rocauc.py [hunk 1]
```diff
@@ -192,6 +197,27 @@ def __init__(
...
+    def fit(self, X, y=None):
+        """
+        Fit the classification model.
+        """
+        # The target determines what kind of estimator is fit
+        ttype = type_of_target(y)
+        if ttype.startswith(MULTICLASS):
+            self.target_type_ = MULTICLASS
+        elif ttype.startswith(BINARY):
+            self.target_type_ = BINARY
```
Comment (bbengfort) [inline yellowbrick/classifier/rocauc.py:209]: I think it's a good idea to store the discovered target type in a learned attribute, but this also needs to be added to the docstring, which will also help users diagnose issues with target types that are are not binary or multiclass.

File: tests/test_classifier/test_rocauc.py [hunk 2]
```diff
@@ -305,42 +343,6 @@ def test_rocauc_no_macro_no_micro(self):
...
-
-    def test_rocauc_no_curves(self):
-        """
-        Test ROCAUC with no curves specified at all
-        """
-        # Create and fit the visualizer
-        visualizer = ROCAUC(
-            LogisticRegression(), per_class=False, macro=False, micro=False
-        )
-        visualizer.fit(self.binary.X.train, self.binary.y.train)
```
Comment (bbengfort) [inline tests/test_classifier/test_rocauc.py:338]: Potentially if we change these tests to use the multiclass dataset instead of the binary dataset will they pass?

File: yellowbrick/classifier/rocauc.py [hunk 3]
```diff
@@ -217,27 +243,14 @@ def score(self, X, y=None):
...
-            if self.micro or self.macro or self.per_class:
+        if self.target_type_ == BINARY:
+            # If it's binary classification, to draw micro or macro curves per_class must be True
+            if (self.micro or self.macro) and not self.per_class:
                 raise ModelError(
-                    "Micro, macro, and per-class scores are not defined for "
-                    "binary classification for estimators with only "
-                    "decision_function methods; set micro, macro, and "
-                    "per-class params to False."
+                    "no curves will be drawn; set per_class=True or micro=False and macro=False."
```
Comment (bbengfort) [inline yellowbrick/classifier/rocauc.py:250]: This error is currently untested.

File: tests/test_classifier/test_rocauc.py (hunk 2 above)
Comment (bbengfort) [reply tests/test_classifier/test_rocauc.py:338]: They do - so I've added the tests back in but using multiclass instead of binary.

File: yellowbrick/classifier/rocauc.py [hunk 4]
```diff
@@ -85,9 +90,17 @@ class ROCAUC(ClassificationScoreVisualizer):
...
+        binary classifiers, setting per_class=False will plot the positive class
+        ROC curve, and per_class=True will use ``1-P(1)`` to compute the curve of
+        the negative class if only a decision_function method exists on the estimator.
+
+    binary : bool, default: False
+        This argument quickly resets the visualizer for true binary classification
+        by updating the micro, macro, and per_class arguments to False (do not use
+        in conjunction with those other arguments). Note that this is not a true
+        hyperparameter to the visualizer, it just collects other parameters into
+        a single, simpler argument.
```
Comment (bbengfort) [inline yellowbrick/classifier/rocauc.py:103]: This metaparameter is just a quick helper for setting micro/macro/per_class to False.

File: yellowbrick/classifier/rocauc.py [hunk 5]
```diff
@@ -254,16 +284,48 @@ def score(self, X, y=None):
...
+        # If the decision is binary draw only ROC curve for the postitive class
+        if self.target_type_ is BINARY and not self.per_class:
+            # In this case predict_proba returns an array of shape (n, 2) which
+            # specifies the probabilities of both the negative and positive classes.
+            if len(y_pred.shape) == 2 and y_pred.shape[1] == 2:
+                self.fpr[BINARY], self.tpr[BINARY], _ = roc_curve(y, y_pred[:,1])
+            else:
+                # decision_function returns array of shape (n,), so plot it directly
+                self.fpr[BINARY], self.tpr[BINARY], _ = roc_curve(y, y_pred)
+            self.roc_auc[BINARY] = auc(self.fpr[BINARY], self.tpr[BINARY])
```
Comment (bbengfort) [inline yellowbrick/classifier/rocauc.py:296]: Using the "binary" label here I think is a bit more informative to the user.
